### DISCUSSION TITLE
How to access the automatically created mac desktop lifetime?

### DISCUSSION THREAD

**Discussion by hhyyrylainen** [AvaloniaUI/Avalonia#14281] [Q&A] [answered]
After updating to Avalonia 11 my app no longer works on mac because something creates a mac lifetime behind the scenes, and I can't find a way to access that so I get an exception when I try to create a lifetime that one already exists. AvaloniaLocator is marked as an internal type and I've looked through all the start variants etc. and I can't find a way to be able to use a lifetime multiple times to run an app  (`StartWithClassicDesktopLifetime` is unusable for me as it disposes stuff like the Dispatcher).

If I tried to reimplement my app GUI restart logic using a lifetime that never quits automatically, I'd need to either poll the active window list or reimplement a custom open window registration system to know when the avalonia side is ready to quit.

So have I overlooked something I could do or is there simply currently not enough powerful internal methods exposed in Avalonia?

Update: it seems that `lifetime.Start(args);` can't be reused anymore at all in Avalonia 11 so I guess I need to reimplement open window detection for my app and switch to that.

**Comment by hhyyrylainen** [comment, marked as answer]
Okay, so I modified my approach to use that custom shutdown logic with just a background task checking every 30 milliseconds how many windows there are open to detect when things should quite (and if things should continue automatically recreating the main window). This is what I did: https://github.com/Revolutionary-Games/Thrive-Launcher/commit/cbe0ee21e0753b1058a8653e4f6442e2cdf4bbd1 
If I can I'll mark this as the answer tomorrow if when testing on a mac these changes work as expected there as well.

**Reply by thevortexcloud** [reply]
I can't really answer anything Mac specific but I feel like you could probably easily solve this by making some sort of window manager, which lets you avoid polling. Anything that wants to open a window would need to go via the manager, which can then subscribe to the close events of the window. This would also give you some better control, such as marking some windows as being ignored for this sort of check. You would also gain an abstraction which makes unit testing much easier since you don't have to worry about real windows suddenly opening.

**Reply by Gillibald** [reply]
https://github.com/AvaloniaUI/Avalonia/pull/14267

**Reply by hhyyrylainen** [reply]
I was considering writing that custom window management code, but then I'd be duplicating logic from the standard runtime types (they have the internal window open and close callbacks that seemingly is not possible for user code to hook into). Just reading the count of the open window list every 30 milliseconds seems to at least work pretty well, and avoids any bugs from forgetting to use the custom window open and close callbacks I would need to write and remember to use for each window.

**Reply by thevortexcloud** [reply]
I have done the manager approach before. It helps if you also make it manage all your view models for you too. EG I typically do something like `windowService.ShowWindowSingleton<MyWindow,MyViewModel>()` which will open either a new window with a new view model instance (or resolve it from the DI container), or bring an existing window that matches the type into the foreground and focuses it. It worked pretty well and easily stopped people opening 500 duplicate windows by accident. Either way, _how_ a window gets opened, as well as their lifecycle is completely abstracted from all my views. It's simply just not up to them. The manager has complete authority over that. Views can still technically open a window directly themselves but that would likely cause other problems.

The other thing it let me do was mark certain windows as being related to each other, which meant I could prevent them from being opened if any of the others were opened.

In any case, it's not for everyone but it certainly has some big advantages if you are willing to stick to using it.

> open and close callbacks that seemingly is not possible for user code to hook into

As I said, there is an open/close event you can use.

https://reference.avaloniaui.net/api/Avalonia.Controls/TopLevel/3DCD51A3
https://reference.avaloniaui.net/api/Avalonia.Controls/TopLevel/82A7C00A

I have used the close event in my window management code to auto dispose my view models if they implement `IDisposable` in the past, and to know when I need to recreate a window if it has been requested to be opened as a singleton.

**Reply by hhyyrylainen** [reply]
Ah yeah, I could use `Window.WindowOpenedEvent.AddClassHandler` to get access to the window events. I'd still need to maintain a custom list of open windows (as relying on my callback being triggered after the one in the lifetime seems brittle). I know my current polling based implementation is not the best, but seems to work and is not complicated at all (as I have just a few windows in my app, I don't think I need to overtly complicate managing them).

**Reply by maxkatz6** [reply]
Please use `Window.WindowOpenedEvent` to track opened/closed windows. Application lifetime has nothing to do with that. 

> Update: it seems that lifetime.Start(args); can't be reused anymore at all in Avalonia 11

You can restart DispatcherFrame/MainLoop, if you need. 
In overall, this whole situation sounds like you don't need application lifetime at all, as it only limits your code. Keep in mind, it's optional.

**Reply by hhyyrylainen** [reply]
I basically need a conceptually pretty simple operation: restart the avalonia app after it has exited due to some condition. This seems to have no inbuilt way to do this simply, so I think the solution of manually keeping a lifetime open until the condition to need to restart can no longer happen seems pretty good.

My usecase is that the app I'm working on is a launcher for another program. If the user closes the launcher window while the launched program is still running I want the launcher process to hang around until the other process ends in order to be able to detect if the other process crashed or ran into another problem that warrants showing a crash reporter window to the user.

**Reply by thevortexcloud** [reply]
> I basically need a conceptually pretty simple operation: restart the avalonia app after it has exited due to some condition. This seems to have no inbuilt way to do this simply

Typically you would do that by starting a separate monitoring/watchdog process. The monitoring process can then decide what to do (such as restart your GUI) if the other process stops. Trying to shove it all into your GUI app feels like you are overcomplicating things to me.

**Reply by hhyyrylainen** [reply]
My avalonia app *is* the watching process that handles the crashes for the process it starts (like I described in my actual use case paragraph). To me setting up one more layer of processes watching each other seems more complicated than the current setup where I have the launcher app (avalonia) and the game it launchers, instead of going C# console app -> avalonia app -> game app (that would complicate things as some state must be preserved between avalonia app launches for it to show correct info).

**Reply by thevortexcloud** [reply]
> My avalonia app is the watching process that handles the crashes for the process it starts 

I get that. 

>  instead of going C# console app -> avalonia app -> game app

My thought process would be the avalonia app would start up a service worker with what ever launch arguments it needs. It can then restart your Avalonia app with some launch argument detailing the error/where to find it. Or you can just have the service worker itself display the error info via Avalonia and then prompt the user if they want to restart the launcher. It's a supported scenario and I have done it myself for some background processing stuff.

https://github.com/AvaloniaUI/Avalonia.Samples/blob/11d74025b25dd468c1388ad962d16b3deaf78a10/src/Avalonia.Samples/MVVM/MsExtensionsHostingSample/README.adoc

Either way, it still feels cleaner to me than forcing your main app to hang around.

**Reply by hhyyrylainen** [reply]
I tested on mac and everything now works fine there as well. I was also able to easily implement the usual way mac apps behave still staying open even without any windows and being able to reactivate.
